### PR TITLE
protocol: add the framework of x protocol, and cmdline arguments.

### PR DESCRIPTION
**Pull request by hicqu**
This is a basic framework for x protocol server side.

X protocol is a new C/S protocol for MySQL from 5.7. It offers document-store access. And It's defined by proto instead of text specifies, so it's more easily to be implemented. **MySQL Shell** and **Mysql Connector** 7.0.2 higher have implemented it already.

### CODE REVIEW

**Comment by hicqu** [comment]
@shenli @tiancaiamao @zimulala  PTAL, thanks.

**Comment by zimulala** [comment]
I don't know what `x protocol server ` is used for. Can you add some explanation?

**Comment by zimulala** [comment]
Please add some tests.

File: tidb-server/main.go [hunk 1]
```diff
@@ -157,6 +167,11 @@ func main() {
 	if err != nil {
 		log.Fatal(errors.ErrorStack(err))
 	}
+	var xsvr *xserver.Server
+	xsvr, err = xserver.NewServer(xcfg)
```
Comment (tiancaiamao) [inline tidb-server/main.go]: Is it necessary to run it by default?

File: util/random.go [hunk 2]
```diff
@@ -0,0 +1,31 @@
+// Copyright 2017 PingCAP, Inc.
```
Comment (tiancaiamao) [inline util/random.go]: You make it a separate package, but not test for it, our coverage down continuous.

**Comment by jackysp** [comment]
Please resolve the conflicts.

File: util/random.go (hunk 2 above)
Comment (hicqu) [inline util/random.go]: It's just a networking level stub. We can only add test cases after we implement them in **business logic level**, just like we do in `server/{sever.go,conn.go}`.

File: tidb-server/main.go (hunk 1 above)
Comment (jackysp) [inline tidb-server/main.go]: Done. PTAL

File: tidb-server/main.go [hunk 3]
```diff
@@ -168,6 +186,7 @@ func main() {
 	go func() {
 		sig := <-sc
 		log.Infof("Got signal [%d] to exit.", sig)
+		xsvr.Close() // Should close xserver before server.
```
Comment (hicqu) [inline tidb-server/main.go]: @jackysp , here we also need to check `xsvr == nil`.

Comment (hicqu) [inline tidb-server/main.go]: rest LGTM.

Comment (jackysp) [inline tidb-server/main.go]: Done.

**Pull request merged**